zone.go

```diff
@@ -115,6 +120,154 @@ func (z *Zone) IsInRootZone() bool {
 	return z.IsTLD() && z.IsDelegated()
 }
 
+// Checks to see if a domain is valid according to character set restriction
+// on the zone.
+// Input must be normalized by the client (lowercase, ASCII-encoded).
+func (z *Zone) IsValidDomain(domain string) bool {
+	if !z.isSubdomain(domain) {
+		return false
+	}
+	if len(z.CodePoints) == 0 {
+		return true
+	}
+	// Don't check the suffix since we've already done that
+	return labelsInCodePoints(z.unicodeLabels(domain), z.CodePoints)
+}
+
+var ErrNotSubdomain = errors.New("domain is not a member of the zone")
+
+// Return the IDN table language the domain matches. Returns an empty string
+// if no IDN tables are defined for the zone or if the domain is not an IDN.
+// Input must be normalized by the client (lowercase, ASCII-encoded).
+func (z *Zone) IDNTable(domain string) (lang string, err error) {
+	if !z.isSubdomain(domain) {
+		return "", ErrNotSubdomain
+	}
+
+	// A blank IDN table language indicates no IDN tables are present
+	if len(z.CodePoints) == 0 {
+		return "", nil
+	}
+
+	// Don't check the suffix since we've already done that, and make sure
+	// we are working with the unicode form, since the ASCII form will never
+	// match an IDN table
+	labels := z.unicodeLabels(domain)
+
+	// Ensure it is an IDN
+	if labelsInCodePoints(labels, asciiCodePoints) {
+		return "", nil
+	}
+
+	if !labelsInCodePoints(labels, z.CodePoints) {
+		return "", errors.New("domain is not a valid member of the zone")
+	}
+
+	for lang, points := range z.IDNTables {
+		if labelsInCodePoints(labels, points) {
+			return lang, nil
+		}
+	}
+
+	return "", errors.New("domain is not a valid IDN of the zone")
+}
+
+func (z *Zone) isSubdomain(domain string) bool {
+	if len(domain) < len(z.Domain)+2 {
+		return false
+	}
+	if domain[len(domain)-len(z.Domain)-1] != '.' {
+		return false
+	}
+	return strings.HasSuffix(domain, z.Domain)
+}
+
+func (z *Zone) unicodeLabels(domain string) []string {
+	prefix, _ := idna.ToUnicode(domain[:len(domain)-len(z.Domain)-1])
+	return strings.Split(prefix, ".")
+}
+
+var asciiCodePoints = []rune{'-', '-', '0', '9', 'a', 'z'}
+
+func labelsInCodePoints(labels []string, points []rune) bool {
+	for _, l := range labels {
+		if !stringInCodePoints(l, points) {
+			return false
+		}
+	}
+	return true
+}
+
+func stringInCodePoints(s string, points []rune) bool {
+	var min rune = '\U0010FFFF'
+	var max rune = 0
+
+	for _, c := range s {
+		if c > max {
+			max = c
+		}
+		if c < min {
+			min = c
+		}
+	}
+
+	if !runeInCodePoints(max, points) {
+		return false
+	}
+	if !runeInCodePoints(min, points) {
+		return false
+	}
+
+	for _, c := range s {
+		if !runeInCodePoints(c, points) {
+			return false
+		}
+	}
+	return true
+}
+
+func runeInCodePoints(c rune, points []rune) bool {
+	return linearRuneInCodePoints(c, points)
+}
+
+func binaryRuneInCodePoints(c rune, points []rune) bool {
```

Comment:
I presume I should remove this implementation and related benchmarks since it is slower than the linear search?